torch/fx/experimental/symbolic_shapes.py

```diff
@@ -374,13 +410,27 @@ def binary_magic_impl(self, other):
             raise
         out = sympy.expand(out)
         pytype: Type
+        # This is not strictly correct. In Python, a**b may return complex when
+        # a < 0 and b is a float: (-1)**2.1. Same for sympy.sqrt(-3.14). This
+        # returns a float while both arguments are ints: 2**(-1). Also, max and
+        # min do not type promote. To avoid having data-dependent control flow
+        # here, we just set the type to float if one of the args is a float. In
+        # case of a type mismatch, we assume that it will be detected during
+        # evaluation.
         if method in always_float_magic_methods:
             pytype = float
+        elif method in ("min", "max"):
+            # This doesn't type-promote in Python, but it's data-dependent, so
+            # we always assume float.
+            pytype = float
```

Comment:
Sorry, so if it doesn't type promote, why is it data dependent?

Reply:
You need to eval to be able to compare, so cannot just rely on the types.